bson/src/main/org/bson/codecs/FlexibleCodec.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs;
+
+import org.bson.BsonType;
+
+/**
+ * Implementations of this interface can decode additional types
+ * and translate them to the desired value type depending on the BsonRepresentation.
+ *
+ * @param <T> the value type
+ * @since 4.2
+ */
+public interface FlexibleCodec<T> extends Codec<T> {
```

Comment:
The motivation for this design is that we need a way to tell Codecs what other type they can read, aside from what they would do normally (we don't want String codecs reading ObjectIds normally)... There's no way to do this currently, so I decided to define a new interface that has a method that allows you to specify what other type it can read.